src/wp-includes/formatting.php

```diff
@@ -1278,32 +1306,271 @@ function wp_check_invalid_utf8( $text, $strip = false ) {
 	if ( ! isset( $is_utf8 ) ) {
 		$is_utf8 = is_utf8_charset();
 	}
-	if ( ! $is_utf8 ) {
+
+	if ( ! $is_utf8 || wp_is_valid_utf8( $text ) ) {
 		return $text;
 	}
 
-	// Check for support for utf8 in the installed PCRE library once and store the result in a static.
-	static $utf8_pcre = null;
-	if ( ! isset( $utf8_pcre ) ) {
-		// phpcs:ignore WordPress.PHP.NoSilencedErrors.Discouraged
-		$utf8_pcre = @preg_match( '/^./u', 'a' );
+	return $strip
+		? wp_scrub_utf8( $text )
+		: '';
+}
+
+/**
+ * Replaces ill-formed UTF-8 byte sequences with the Unicode Replacement Character.
+ *
+ * While it’s usually safe to ignore invalid UTF-8, there are cases where it’s necessary
+ * to work only with a valid string, such as when producing XML or feeding data into a
+ * large language model. In these cases, calling this function will produce such a valid
+ * string where the invalid segments are replaced with the Unicode Replacement Character.
```

Comment:
This tone makes it seem like invalid UTF-8 is usually fine, but I'm reluctant to give that impression. I'd prefer less relaxed language:

```suggestion
 * It's possible to ignore invalid UTF-8, but there are cases where it’s necessary
 * to work only with a valid string, such as when producing XML or feeding data into a
 * large language model. In these cases, calling this function will produce such a valid
 * string where the invalid segments are replaced with the Unicode Replacement Character.
```

Reply:
After thinking about your comment I have changed the wording considerably. I welcome your thoughts on the new construction. I’m trying to balance passing through invalid bytes for round-tripping value and scrubbing to avoid safety issues.

It’s really hard to come up with cases where we _should_ scrub. The more I look, the fewer places I find. The only places I can think of are where some external interface crashes in the presence of invalid bytes, such as XML. For every other case I think there are strong arguments to be made about leaving the “bad” bytes in place.